pynicotine/portmapper.py

```diff
@@ -126,17 +127,23 @@ def _request_port_mapping(self, public_port, private_port, lease_duration):
 
         with socket.socket(family=socket.AF_INET, type=socket.SOCK_DGRAM, proto=socket.IPPROTO_UDP) as sock:
             sock.bind((self.local_ip_address, 0))
-            sock.settimeout(self.REQUEST_TIMEOUT)
-
             request = self.PortmapRequest(public_port, private_port, lease_duration)
-            request.sendto(sock, (self._gateway_address, self.REQUEST_PORT))
+            timeout = self.REQUEST_INIT_TIMEOUT
+            i = 1
 
-            try:
-                response = self.PortmapResponse(message=sock.recv(16))
-                return response.result
+            while i <= self.REQUEST_ATTEMPTS:
+                sock.settimeout(timeout)
+                request.sendto(sock, (self._gateway_address, self.REQUEST_PORT), i)
```

Comment:
I don't think the internal parenthesis are necessary.

Reply:
Do you mean `(self._gateway_address, self.REQUEST_PORT)`? How would you indicate that it's a tuple then? I tried removing them and the interpreter considered the tuple parts as separate function arguments.

Reply:
Our sendto() function does not require a tuple,
it takes 4 arguments `(self, sock, addr, num_attempt)`.

Reply:
In any case, this is pre-existing code that affects UPnP as well. It's not relevant to this PR.

Reply:
It doesn't really make any difference either way.

My mistake I missed that `addr` is in fact a tuple of `(addr, port)`